test/test_test.c

```diff
@@ -22,28 +22,34 @@
 #include "test_main.h"
 #include "testutil.h"
 
-#define C(l, b, t)                                      \
-    if ((t) != b) {                                     \
-        fprintf(stderr, "FATAL : %s != %d\n", #t, b);   \
-        goto l;                                         \
+#define TEST(b, t) test_case(#t, (t), (b))
+
+static int test_case(const char *test, int result, int expected)
+{
+    if (result != expected) {
+        fprintf(stderr, "FATAL: %s != %d\n", test, expected);
+        return 0;
     }
+    return 1;
+}
 
 static int test_int(void)
 {
-    C(err, 1, TEST_int_eq(1, 1));
-    C(err, 0, TEST_int_eq(1, -1));
-    C(err, 1, TEST_int_ne(1, 2));
-    C(err, 0, TEST_int_ne(3, 3));
-    C(err, 1, TEST_int_lt(4, 9));
-    C(err, 0, TEST_int_lt(9, 4));
-    C(err, 1, TEST_int_le(4, 9));
-    C(err, 1, TEST_int_le(5, 5));
-    C(err, 0, TEST_int_le(9, 4));
-    C(err, 1, TEST_int_gt(8, 5));
-    C(err, 0, TEST_int_gt(5, 8));
-    C(err, 1, TEST_int_ge(8, 5));
-    C(err, 1, TEST_int_ge(6, 6));
-    C(err, 0, TEST_int_ge(5, 8));
+    if (!TEST(1, TEST_int_eq(1, 1))
+        && !TEST(0, TEST_int_eq(1, -1))
+        && !TEST(1, TEST_int_ne(1, 2))
+        && !TEST(0, TEST_int_ne(3, 3))
+        && !TEST(1, TEST_int_lt(4, 9))
+        && !TEST(0, TEST_int_lt(9, 4))
+        && !TEST(1, TEST_int_le(4, 9))
+        && !TEST(1, TEST_int_le(5, 5))
+        && !TEST(0, TEST_int_le(9, 4))
+        && !TEST(1, TEST_int_gt(8, 5))
+        && !TEST(0, TEST_int_gt(5, 8))
+        && !TEST(1, TEST_int_ge(8, 5))
+        && !TEST(1, TEST_int_ge(6, 6))
+        && !TEST(0, TEST_int_ge(5, 8)))
+        goto err;
```

Comment:
I think you need to rethink the semantics here...  With this condition, this function will succeed if *one* of the test cases succeed.

So general rule of thumb, you should write these cascading conditions like this:

``` C
    if (!cond1
        || !cond2
        || !cond3)
        /* failure, do whatever to signal that, like goto err; */
```

or:

``` C
    if (cond1
        && cond2
        && cond3)
        /* success, do whatever to signal that, like return 1; */
```

Makes sense?